request/fetch.go

```diff
@@ -0,0 +1,140 @@
+			"The requested password file does not have the expected '.gpg' extension",
+			&map[string]string{"action": "fetch", "file": request.File},
+		)
+	}
+
+	store, ok := request.Settings.Stores[request.Store]
+	if !ok {
+		log.Errorf(
+			"The password store '%v' is not present in the list of stores '%v'",
+			request.Store, request.Settings.Stores,
+		)
+		response.SendErrorAndExit(
+			errors.CodeInvalidPasswordStore,
+			"The password store is not present in the list of stores",
+			&map[string]string{"action": "fetch", "name": request.Store},
+		)
+	}
+
+	normalizedStorePath, err := normalizePasswordStorePath(store.Path)
+	if err != nil {
+		log.Errorf(
+			"The password store '%v' is not accessible at the location '%v': %+v",
+			store.Name, store.Path, err,
+		)
+		response.SendErrorAndExit(
+			errors.CodeInaccessiblePasswordStore,
+			"The password store is not accessible",
+			&map[string]string{"action": "fetch", "error": err.Error(), "name": store.Name, "path": store.Path},
+		)
+	}
+	store.Path = normalizedStorePath
+
+	gpgPath := request.Settings.GpgPath
+	if gpgPath != "" {
+		err = validateGpgBinary(gpgPath)
+		if err != nil {
+			log.Errorf(
+				"The provided gpg binary path '%v' is invalid: %+v",
+				gpgPath, err,
+			)
+			response.SendErrorAndExit(
+				errors.CodeInvalidGpgPath,
+				"The provided gpg binary path is invalid",
+				&map[string]string{"action": "fetch", "error": err.Error(), "gpgPath": gpgPath},
+			)
+		}
+	} else {
+		gpgPath, err = detectGpgBinary()
+		if err != nil {
+			log.Error("Unable to detect the location of the gpg binary: ", err)
+			response.SendErrorAndExit(
+				errors.CodeUnableToDetectGpgPath,
+				"Unable to detect the location of the gpg binary",
+				&map[string]string{"action": "fetch", "error": err.Error()},
+			)
+		}
+	}
+
+	responseData.Contents, err = decryptFile(store, request.File, gpgPath)
+	if err != nil {
+		log.Errorf(
+			"Unable to decrypt the password file '%v' in the password store '%v' located in '%v': %+v",
+			request.File, store.Name, store.Path, err,
+		)
+		response.SendErrorAndExit(
+			errors.CodeUnableToDecryptPasswordFile,
+			"Unable to decrypt the password file",
+			&map[string]string{"action": "fetch", "error": err.Error(), "file": request.File, "name": store.Name, "path": store.Path},
+		)
+	}
+
+	response.SendOk(responseData)
+}
+
+func detectGpgBinary() (string, error) {
+	// Look in $PATH first, then check common locations - the first successful result wins
+	gpgBinaryPriorityList := []string{
+		"gpg2", "gpg",
+		"/bin/gpg2", "/usr/bin/gpg2", "/usr/local/bin/gpg2",
+		"/bin/gpg", "/usr/bin/gpg", "/usr/local/bin/gpg",
+	}
+
+	for _, binary := range gpgBinaryPriorityList {
+		err := validateGpgBinary(binary)
+		if err == nil {
+			return binary, nil
+		}
+	}
+
+	return "", goerrors.New("Unable to detect the location of the gpg binary to use")
+}
+
+func validateGpgBinary(gpgPath string) error {
+	return exec.Command(gpgPath, "--version").Run()
```

Comment:
Is this just to check that it's possible to execute the binary, rather than actually getting any specific info from it?

Reply:
Exactly, a cheap way to test whether a user didn't make a mistake when configuring `gpgPath`. Or if they didn't configure this setting at all, I'll enumerate some "popular" locations for where the `gpg` might exist and use this function to test if any of those locations contain the `gpg` binary that we can use.